runtime/src/main/jni/ObjectManager.cpp

```diff
@@ -263,6 +286,56 @@ void ObjectManager::JSObjectWeakCallbackStatic(const WeakCallbackInfo<ObjectWeak
     thisPtr->JSObjectWeakCallback(isolate, callbackState);
 }
 
+void ObjectManager::JSObjectFinalizerStatic(const WeakCallbackInfo<ObjectWeakCallbackState>& data) {
+    ObjectWeakCallbackState* callbackState = data.GetParameter();
+
+    ObjectManager* thisPtr = callbackState->thisPtr;
+
+    auto isolate = data.GetIsolate();
+
+    thisPtr->JSObjectFinalizer(isolate, callbackState);
+}
+
+void ObjectManager::JSObjectFinalizer(Isolate* isolate, ObjectWeakCallbackState* callbackState) {
+    Persistent<Object>* po = callbackState->target;
+
+    auto jsInfoIdx = static_cast<int>(MetadataNodeKeys::JsInfo);
+    auto jsInstance = po->Get(m_isolate);
+    auto jsInfo = jsInstance->GetInternalField(jsInfoIdx);
+    if (jsInfo->IsUndefined()) {
+        // Typescript object layout has an object instance as child of the actual registered instance. checking for that
+        auto prototypeObject = jsInstance->GetPrototype().As<Object>();
+        if (!prototypeObject.IsEmpty() && prototypeObject->IsObject()) {
+            DEBUG_WRITE("GetJSInstanceInfo: need to check prototype :%d", prototypeObject->GetIdentityHash());
+            if (IsJsRuntimeObject(prototypeObject)) {
+                jsInfo = prototypeObject->GetInternalField(jsInfoIdx);
+            }
+        }
+    }
+
+    if (jsInfo.IsEmpty() || !jsInfo->IsExternal()) {
+        // The JavaScript instance has been forcefully disconnected from the Java instance.
```

Comment:
We may add explicit "release()" method in future (although highly unlikely). But it also does some sanity check like for the preconditions that we have to check and you mention two lines below.